### PR TITLE
GearmanCollector: Emit task specific queue length and worker counts

### PR DESCRIPTION
**Pull request by venkey-ariv**

### CODE REVIEW

File: src/diamond/collectors/gearman_stats/gearman_stats.py [hunk 1]
```diff
@@ -60,6 +54,23 @@ def get_fds(gearman_pid_path):
             proc_path = os.path.join('/proc', gearman_pid, 'fd')
             return len(os.listdir(proc_path))
 
+        def publish_server_stats(gm_admin_client):
+            #  Publish idle/running worker counts
+            #  and no. of tasks queued per task
+            for entry in gm_admin_client.get_status():
+                total = entry['workers']
+                running = entry['running']
+                idle = total-running
```
Comment (tsheasha) [inline src/diamond/collectors/gearman_stats/gearman_stats.py]: are you sure these are returned as int and not str?

File: src/diamond/collectors/gearman_stats/gearman_stats.py [hunk 2]
```diff
@@ -60,6 +54,23 @@ def get_fds(gearman_pid_path):
...
+
+                self.dimensions = {'task': entry['task']} # Internally, this dict is cleared on self.publish
+                self.publish('gearman.queued', entry['queued'])
+
+                self.dimensions = {'type': 'running'}
+                self.publish('gearman.workers', running)
+
+                self.dimensions = {'type': 'idle'}
+                self.publish('gearman.workers', idle)
+
```
Comment (tsheasha) [inline src/diamond/collectors/gearman_stats/gearman_stats.py]: +1 this is a very positive change!

File: src/diamond/collectors/gearman_stats/gearman_stats.py [hunk 3]
```diff
@@ -60,6 +54,23 @@ def get_fds(gearman_pid_path):
             proc_path = os.path.join('/proc', gearman_pid, 'fd')
             return len(os.listdir(proc_path))
 
+        def publish_server_stats(gm_admin_client):
+            #  Publish idle/running worker counts
+            #  and no. of tasks queued per task
+            for entry in gm_admin_client.get_status():
+                total = entry['workers']
+                running = entry['running']
```
Comment (tsheasha) [inline src/diamond/collectors/gearman_stats/gearman_stats.py]: maybe for these you want running = entry.get('running', 0) and total = entry.get('workers', 0)

File: src/diamond/collectors/gearman_stats/gearman_stats.py (hunk 1 above)
Comment (tsheasha) [inline src/diamond/collectors/gearman_stats/gearman_stats.py]: also if you use the approach i mentioned baove of .get("key", 0) please make sure to wrap this in a abs() to get the absolute value.

**Comment by tsheasha** [comment]
fix and +1

**Comment by gnufied** [comment]
:shipit:

File: src/diamond/collectors/gearman_stats/gearman_stats.py (hunk 3 above)
Comment (venkey-ariv) [inline src/diamond/collectors/gearman_stats/gearman_stats.py]: Done

File: src/diamond/collectors/gearman_stats/gearman_stats.py (hunk 1 above)
Comment (venkey-ariv) [inline src/diamond/collectors/gearman_stats/gearman_stats.py]: Only absolute numbers are returned and they seem to integers :)

**Pull request merged**